### PR TITLE
test(fixer): cover more test cases

### PR DESCRIPTION
**Pull request by jason89521**
## Description

Almost all ESLint's test cases are ported. However, there are some test case cannot be satisfied:

1. BOM removal: ESLint use `range.start===-1` to determine whether to delete the starting BOM, but `Span` cannot accept negative number.
2. Messages sorting: If there are multiple nofix messages, eslint we sort them by there column and line.

## Related issue

#68

### CODE REVIEW

**Comment by jason89521** [comment]
For BOM removal, we can add a flag property in `Fix`. Is this approach appropriate?

**Comment by jason89521** [comment]
For messages sorting, we can add `Span` to `Message` struct.

**Comment by Boshen** [comment]
Both are interesting problems, let me think about these.

**Comment by Boshen** [comment]
For BOM, it seems our parser will parse it and offset all the spans correctly, can you test this and see if this is correct?

**Comment by Boshen** [comment]
What is message sorting exactly? Can you give an example?

**Comment by jason89521** [comment]
> For BOM, it seems our parser will parse it and offset all the spans correctly, can you test this and see if this is correct?

I prepend a `\uFEFF` to the test.js, and the parser said that it is an invalid character. Not sure whether the way to check this is correct or not.
image

**Comment by jason89521** [comment]
> What is message sorting exactly? Can you give an example?

Suppose we have two messages:

```
NoFix1 {
  error: no_fix_error1,
  line: 1,
  column: 0
  fix: None,
  fixed: false,
}

NoFix2 {
  error: no_fix_error2,
  line: 1,
  column: 10
  fix: None,
  fixed: false
}
```

Since they cannot be fixed, the fixer will consider them remaining unfixed message. For ESLint, they will sort them by the `line` and `column`. In this example, `NoFix1` will be in front of `NoFix2`.

**Comment by Boshen** [comment]
```
oxc  main ❯ cargo run -p oxc_parser --example parser -- ./tasks/coverage/typescript/tests/cases/compiler/bom-utf8.ts
    Finished dev [unoptimized + debuginfo] target(s) in 0.35s
     Running `target/debug/examples/parser ./tasks/coverage/typescript/tests/cases/compiler/bom-utf8.ts`
{
  "type": "Program",
  "start": 0,
  "end": 14,
  "sourceType": "module",
  "body": [
    {
      "type": "VariableDeclaration",
      "start": 3, <-------------------------------- starts at 3
      "end": 12,
      "kind": "var",
      "declarations": [
```

I think eslint had to do this to be compatible with different parsers, which we don't need to.

**Comment by Boshen** [comment]
> For ESLint, they will sort them by the line and column. In this example, NoFix1 will be in front of NoFix2.

Does this mean all we have to do is sort by span.start?

**Comment by Boshen** [comment]
Why does it need to sort? For printing unfixed messages in order?

**Comment by jason89521** [comment]
I don't kwon why eslint sort the unfixed messages, I think it is a bit unnecessary. How do you think?

**Comment by jason89521** [comment]
Does the order of the unfixed message really matter?

**Comment by Boshen** [comment]
> 

It's more readable and user friendly.

**Comment by jason89521** [comment]
> So the conclusion here is that we can drop the BOM test cases? 

Ya, let me remove them.

**Comment by Boshen** [comment]
Can you write some comments on this in the place where you deleted the tests so the next person knows BOM handling is not implemented on purpose.

File: crates/oxc_linter/src/autofix/fixer.rs [hunk 1]
```diff
@@ -356,170 +356,9 @@ mod test {
         ]);
         assert_eq!(result.fixed_code, TEST_CODE.replace("answer", "foo"));
         assert_eq!(result.messages.len(), 2);
+        // We should sort the error to pass the following assertion.
```
Comment (Boshen) [inline crates/oxc_linter/src/autofix/fixer.rs:359]: What does this comment mean?

Comment (Boshen) [reply crates/oxc_linter/src/autofix/fixer.rs:359]: Can you ignore this if it's not passing write now.

Comment (Boshen) [reply crates/oxc_linter/src/autofix/fixer.rs:359]: Oh, you should add a `TODO:` here.

Comment (jason89521) [reply crates/oxc_linter/src/autofix/fixer.rs:359]: It is already ignored. It means that we need to sort the remaining unfixed messages.

Comment (Boshen) [reply crates/oxc_linter/src/autofix/fixer.rs:361]: Yes if it's easy to do so, let's do another PR for this after this.

**Pull request merged**